### ISSUE TITLE
[BUG] Raspberry PI - log-flooding / intermittent failures

### ISSUE THREAD

**Issue by magicdude4eva** [cagnulein/qdomyos-zwift#2954]
**Describe the bug**
This is a follow-up from https://github.com/cagnulein/qdomyos-zwift/discussions/2939 and the Pi installation as described in https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/10_Installation.md

I noticed that the QZ service randomly stops this morning and I do not see anything in the log-files. Restarting the service seems to work again. When the treadmill is off, QZ in debug mode logs a lot of verbose data, which makes the journal unreadable and seems to log batches of messages like this:
[qz-log-debug-redundant.txt.zip](https://github.com/user-attachments/files/18268943/qz-log-debug-redundant.txt.zip)

For context: I automatically restart QZ via crontab at 5am in the morning: `0 5 * * * systemctl restart qz`. This morning I did a treadmill run at 7:46 and Zwift recorded fine via QZ. At around 9:30 I wanted to try something else and noticed that Zwift did not show QZ. Checking the log via `journalctl -u qz --since "2024-12-29 07:45:00" --until "2024-12-29 09:30:00"` did not show anything meaningful other than flooded logs. Restarting the service via `systemctl restart qz` made the threadmill discoverable again.

How would I be able to identify the cause of this "disconnect", especially since QZ worked some 2 hours before?

**To Reproduce**
1. Install the service as described in the Wiki
2. Run the service and then disconnect the treadmill
3. Observe the flooding of logs

**Expected behavior**
If a previously connected device is turned off, QZ should not flood debug logs or provide an alternative info/error/warn logging mechanism to make it easier to find issues such as service failures.

**Screenshots**
The service is started with logging enabled:
<img width="1914" alt="image" src="https://github.com/user-attachments/assets/d9302148-fede-4813-9f11-8f5e687bc61e" />

Service file:
<img width="519" alt="image" src="https://github.com/user-attachments/assets/17a4d295-6a8e-464a-b867-7342f0538b61" />

**Desktop (please complete the following information):**
 - Linux KICKR-RUN 6.6.67-v8+ 1833 SMP PREEMPT Fri Dec 20 14:14:15 GMT 2024 aarch64 GNU/Linux
 - Raspberry Pi 4 Model B Rev 1.5

**Smartphone (please complete the following information):**

**Append a debug log**
I could not export a full service log as it was flushed :-( I am not sure if there is an easier way to filter out relevant information via `journalctl -u qz --since "today" | grep -i ????`
  
  
**Additional context**

**Comment by cagnulein** [comment]
was the treadmill power on the whole time? consider that raspberry cached the bluetooth devices, so if you leave qz running the whole time, it tries to connect to a treadmill that actually it's not there. i think this is the issue. why don't you start the raspberry along with the treadmill?

**Comment by cagnulein** [comment]
@ziozanna @nix155  could help on this too since they have already a lot of experience on this context

**Comment by nix155** [comment]
If you want to configure the service to restart after it stops responding to requests (hangs), you need a script that will run via cron, check the service status, and restart it if the check fails. The remaining task is to define the health/unhealthy criteria.

**Comment by nix155** [comment]
What can be done is to check the application log—look for the messages that are logged when the simulator is found and when it is lost or disconnected (perhaps @cagnulein can provide some insight on this). After that, it's not difficult to write a script: we read the log, search for messages about the loss of the simulator, count the occurrences, and if the count exceeds `n`, we restart the service/clear the log.

**Comment by cagnulein** [comment]
Yeah I guess the "operation already in progress" is a good trigger for this

**Comment by magicdude4eva** [comment]
The automatic restarting would not be really a problem and to script this. I don't think "Operation already in progress" is a good indicator, as with the treadmill currently off, I get 2660 log messages alone between 13:58:50 and 13:20:17:
<img width="1472" alt="image" src="https://github.com/user-attachments/assets/c605e8bd-afc2-4834-b19c-894e32c7f256" />

Hence my concern that logging is a bit too verbose in Debug to keep debug on and there seems to be no log-level control (i.e. --log-level INFO) - or?

**Comment by cagnulein** [comment]
@magicdude4eva Can't you simply use Bluetoothctl to bring up qz when the treadmill is on and close it when it's off. It seems elegant to me

**Comment by magicdude4eva** [comment]
> @magicdude4eva Can't you simply use Bluetoothctl to bring up qz when the treadmill is on and close it when it's off. It seems elegant to me

that is fairly easy, i was hoping that QZ would do this connection management if run on a raspberry/linux in headless mode? if not, a script is fairly easy:
```
#!/bin/bash

TREADMILL_ADDRESS="CA:DD:78:BF:26:0A"

# Check the treadmill's connection status
connected=$(bluetoothctl info "$TREADMILL_ADDRESS" | grep -i "Connected: yes")

if [[ -n "$connected" ]]; then
    echo "Treadmill is ON and connected."
    # Start QZ or other relevant actions
    systemctl is-active --quiet qz || systemctl restart qz
else
    echo "Treadmill is OFF or disconnected."
    # Stop QZ or other relevant actions
    systemctl is-active --quiet qz && systemctl stop qz
fi

```

**Comment by cagnulein** [comment]
Nope as I said qz from qt has no way to know if a device is alive because qt is caching it. It would require a full discovery that takes time

So your script is perfect for this

**Comment by cagnulein** [comment]
You wrote connected but you are not actually connected. It's just advertising

**Comment by magicdude4eva** [comment]
> You wrote connected but you are not actually connected. It's just advertising

This is correct—thanks for clarifying.

However, I have not found a reliable way to determine whether the treadmill is on or off. It seems the closest approach is to periodically attempt a connection using:

```
bluetoothctl connect CA:DD:78:BF:26:0A
```

If the Raspberry Pi can successfully connect to the treadmill, it would start the qz service. If the connection attempt fails and the qz service is running, it would stop the service.

Additionally, I suspect there may be a race condition in the qz service when the treadmill is turned off. In the logs, around the time of a disconnection, I see the following entries:

```
Dec 29 14:20:55 KICKR-RUN qdomyos-zwift[40014]: Sun Dec 29 14:20:55 2024 1735478455487 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString&) "BTLE stateChanged InvalidService".
```
Eventually, the logs flood with entries like:

```
QLowEnergyControllerPrivate setting state to QLowEnergyController::UnconnectedState
...
...
QLowEnergyControllerPrivate setting state to QLowEnergyController::ConnectingState
```

This behavior might lead to a race condition. In my understanding (and I may be oversimplifying), the qz service should, after a device disconnect, reset to the same state it was in when it started up.

If @cagnulein or @nix155 have ideas on how to reliably detect when the treadmill is turned on or off, I could script this better. Below is my current implementation idea (and then run it as a systemd timer every 30 seconds):

/root/qz-treadmill-monitor.sh
```
#!/bin/bash

# Configuration
TREADMILL_MAC="CA:DD:78:BF:26:0A"
QZ_SERVICE="qz.service"
LOG_FILE="/var/log/qz-treadmill-monitor.log"

# Log function
log() {
    echo "$(date '+%Y-%m-%d %H:%M:%S') $1" >> "$LOG_FILE"
}

# Check treadmill status
bluetoothctl connect "$TREADMILL_MAC" &> /dev/null
if [ $? -eq 0 ]; then
    # Successfully connected
    log "Treadmill is ON. Starting $QZ_SERVICE."
    systemctl is-active --quiet "$QZ_SERVICE" || systemctl start "$QZ_SERVICE"
    bluetoothctl disconnect "$TREADMILL_MAC" &> /dev/null
else
    # Failed to connect
    log "Treadmill is OFF. Stopping $QZ_SERVICE."
    systemctl is-active --quiet "$QZ_SERVICE" && systemctl stop "$QZ_SERVICE"
fi

```

**Comment by cagnulein** [comment]
> This behavior might lead to a race condition. In my understanding (and I may be oversimplifying), the qz service should, after a device disconnect, reset to the same state it was in when it started up.

it does and it indeed go to unconnected state to the connecting one over and over, because, as I already said, the linux bluetooth stack is always saying that a device is alive once it's connected. so that's why that you need an external trigger to reset it (or simply turn it on).

I can't check your log because I don't have a linux system here (I'm in vacation), i don't know if a connection as you did would work, maybe from bluetoothctl you can check the timestamp of the last advertisment (i really don't know, i'm just guessing).

If you find a way let me know.

The method that i always suggest is to put a simply power outlet that gives power to the raspberry and the treadmill together. easy and fast :)

**Comment by nix155** [comment]
I don't like the idea of keeping QZ with Bluetooth always on either, especially since the Bluetooth on the simulator itself turns off quite quickly when not in use. 

Writing a Python script to check the availability of the simulator via Bluetooth is straightforward, but the solution proposed by @cagnulein (simultaneous turning on and off) seems to me the most elegant.

**Comment by nix155** [comment]
Here are examples of scripts (functionality not tested):
### Example for Classic Bluetooth with PyBluez

```python
import bluetooth

def connect_to_device(target_name):
    print("Scanning Bluetooth devices...")
    devices = bluetooth.discover_devices(duration=8, lookup_names=True)

    target_address = None
    for addr, name in devices:
        print(f"Found device: {name} ({addr})")
        if target_name.lower() in name.lower():
            target_address = addr
            print(f"Device '{target_name}' found with address {addr}.")
            break

    if target_address is None:
        print(f"Device '{target_name}' not found.")
        return False

    # Establishing a connection
    try:
        print(f"Connecting to device {target_address}...")
        sock = bluetooth.BluetoothSocket(bluetooth.RFCOMM)
        sock.connect((target_address, 1))  # 1 - standard RFCOMM port
        print("Connection established!")
        sock.close()
        return True
    except bluetooth.BluetoothError as e:
        print(f"Connection error: {e}")
        return False

if __name__ == "__main__":
    target_device = "Your trainer's name"
    if connect_to_device(target_device):
        print("Device is available.")
    else:
        print("Failed to connect to the device.")
```

---

### Example for BLE with Bleak

```python
import asyncio
from bleak import BleakScanner, BleakClient

async def connect_to_device(target_name):
    print("Scanning BLE devices...")
    devices = await BleakScanner.discover()

    target_device = None
    for device in devices:
        print(f"Found device: {device.name} ({device.address})")
        if target_name.lower() in (device.name or "").lower():
            target_device = device
            print(f"Device '{target_name}' found with address {device.address}.")
            break

    if target_device is None:
        print(f"Device '{target_name}' not found.")
        return False

    # Establishing a connection
    try:
        async with BleakClient(target_device.address) as client:
            print(f"Connecting to device {target_device.address}...")
            is_connected = await client.is_connected()
            if is_connected:
                print("Connection established!")
                # Display device characteristics
                for service in client.services:
                    print(f"Service: {service.uuid}")
                    for char in service.characteristics:
                        print(f"  Characteristic: {char.uuid}")
            return True
    except Exception as e:
        print(f"Connection error: {e}")
        return False

if __name__ == "__main__":
    target_device = "Your trainer's name"
    asyncio.run(connect_to_device(target_device))
```

**Comment by magicdude4eva** [comment]
After testing the suggested Python script, I encountered issues where it failed to discover devices reliably, even when the treadmill was actively broadcasting. For my use case, Python added unnecessary overhead, and bluetoothctl has proven to be a more reliable tool for detecting the treadmill's status.

I created a Bash script (qz-treadmill-monitor.sh) that utilizes bluetoothctl to periodically check the availability of the treadmill. The script performs the following actions:

1. Discovery Check: It scans for devices and looks for the treadmill's Bluetooth name (I_TL in my case).
2. Service Management: If the treadmill is detected, the script starts the QZ service. If not, it stops the service.
3. Logging: Logs are written to /var/log/qz-treadmill-monitor.log for tracking and debugging.

The core script `/root/qz-treadmill-monitor.sh`
```
#!/bin/bash

LOG_FILE="/var/log/qz-treadmill-monitor.log"
TARGET_DEVICE="I_TL"
SCAN_INTERVAL=30  # Time in seconds between checks
SERVICE_NAME="qz"

log() {
    echo "$(date '+%Y-%m-%d %H:%M:%S') - $1" >> "$LOG_FILE"
}

is_service_running() {
    systemctl is-active --quiet "$SERVICE_NAME"
    return $?
}

scan_for_device() {
    log "Starting Bluetooth scan for $TARGET_DEVICE..."
    
    # Run bluetoothctl scan in the background and capture output
    bluetoothctl scan on &>/dev/null &
    SCAN_PID=$!

    # Allow some time for devices to appear
    sleep 5

    # Check if the target device appears in the list
    bluetoothctl devices | grep -q "$TARGET_DEVICE"
    DEVICE_FOUND=$?

    # Stop scanning
    kill "$SCAN_PID"
    bluetoothctl scan off &>/dev/null

    if [ $DEVICE_FOUND -eq 0 ]; then
        log "Device '$TARGET_DEVICE' found."
        return 0
    else
        log "Device '$TARGET_DEVICE' not found."
        return 1
    fi
}

manage_service() {
    local device_found=$1
    if $device_found; then
        if ! is_service_running; then
            log "Starting QZ service..."
            systemctl start "$SERVICE_NAME"
        else
            log "QZ service is already running."
        fi
    else
        if is_service_running; then
            log "Stopping QZ service..."
            systemctl stop "$SERVICE_NAME"
        else
            log "QZ service is already stopped."
        fi
    fi
}

while true; do
    log "Checking for treadmill status..."
    if scan_for_device; then
        manage_service true
    else
        manage_service false
    fi
    log "Waiting for $SCAN_INTERVAL seconds before next check..."
    sleep "$SCAN_INTERVAL"
done
```

I then created a service file `/etc/systemd/system/qz-treadmill-monitor.service`
```
[Unit]
Description=QZ Treadmill Monitor Service
After=bluetooth.service

[Service]
Type=simple
ExecStart=/root/qz-treadmill-monitor.sh
Restart=always
RestartSec=10
User=root

[Install]
WantedBy=multi-user.target

```

And then activated:
```
sudo systemctl daemon-reload
sudo systemctl enable qz-treadmill-monitor
sudo systemctl start qz-treadmill-monitor

```

The Bash-based approach using bluetoothctl has been more reliable for me compared to Python. This setup ensures the treadmill's status is monitored, and the QZ service starts/stops appropriately. If anyone has ideas to further improve the process or detect the treadmill more efficiently, I’m happy to test and collaborate.

Example of logging when switching treadmill on/off - I did notice that bluetoothctl will take sometimes 30 seconds to "unsee" a turned off device:
<img width="793" alt="image" src="https://github.com/user-attachments/assets/91bfd527-33c9-4a40-ba0d-c4826d97a599" />

**Comment by magicdude4eva** [comment]
Issue can be closed, I have created PR #2960  to adjust documentation
